### PR TITLE
bpo-19903: IDLE: Calltips changed to use inspect.signature

### PR DESCRIPTION
**Pull request by mlouielu**
This commit change the get_argspec from using inspect.getfullargspec
to inspect.signature. It will improve the tip message for use.

Also, if object is not callable, now will return this message for
user, not blank tips. If the methods has an invalid method signature,
it will also return message to user.

Re-open from #1382

https://bugs.python.org/issue19903

### CODE REVIEW

**Comment by terryjreedy** [comment]
Louie, I cannot add commits to this issue, and a few others of yours and others.  Please follow the instructions at https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/#enabling-repository-maintainer-permissions-on-existing-pull-requests to make sure "Allow edits from maintainers." is checked for this and other IDLE issues.  Please let me know what you found.  If that is already checked, then I have some other problem with github.

**Comment by mlouielu** [comment]
@terryjreedy The check is on now. Hmmm, I thought that is default to check-up? FWIR this has been discussed on python-committers before?

**Comment by terryjreedy** [comment]
There seems to have been a problem with a robot unchecking the edit switch under some circumstances.  Since patched.

**Comment by terryjreedy** [comment]
I want to make a few minor changes, do some live tests, and merge.  But when I tried to make a local PR, git trapped my clone in a loop.  ~~See core-mentorship.~~ Fixed

File: Lib/idlelib/calltips.py
Comment (terryjreedy) [inline Lib/idlelib/calltips.py]: /argspec = ""/argspec = default = ""/ above
/argspec/default/ here
-- done

File: Lib/idlelib/calltips.py
Comment (terryjreedy) [inline Lib/idlelib/calltips.py]: Replace with conditional expression.
-- done

File: Lib/idlelib/calltips.py
Comment (terryjreedy) [inline Lib/idlelib/calltips.py]: Replace body with conditional expression:
return _invalid_method if str(err).startswith(_invalid_method) else default

**Review by terryjreedy — CHANGES_REQUESTED** [review]
I wrote minor changes I planned to write, to make code slightly clearer.  The should not affect tests.

PR also needs a blurb added.  The title line might be enough.  In any case, web edit works well for blurbs.

File: Lib/idlelib/calltips.py
Comment (mlouielu) [reply Lib/idlelib/calltips.py]: If we do this, it will break the unittest, for e.g. `range` will not able to return argspec.

Comment (terryjreedy) [reply Lib/idlelib/calltips.py]: Right, in these cases, we still look for docstring.

**Review by terryjreedy — APPROVED** [review]
I am doing some manual checking while waiting for CI, but expect to merge.

**Comment by mlouielu** [comment]
@terryjreedy Can this be backported to 3.5 and 3.6?

**Comment by terryjreedy** [comment]
Funny you should ask.  When 3.6.0b1 came out, I stopped backporting to anything before 3.6: too much work and too much danger of regressions.  In testing this, I noticed that ')' no longer closes calltips  -- a bad regression since 3.6.2.  Since repository 3.6.2+ has the same problem, it is definitely not this patch.  I suspect bpo-30723.  Of course, there is no test of such behaviors.

That aside, this patch would have less effect in 3.5 (pre-Arg Clinic, I believe), and the
last 3.5 maintenance release candidate is already out, so it is effectively in security-fix only mode.


too dangerousCurrently, everything is backported to 3.6, and only 3.6.  The last 3.5 maintenance release candidate was issued a week ago, so it is effectively in security-only mode.

File: Lib/idlelib/calltips.py
Comment (terryjreedy) [inline Lib/idlelib/calltips.py]: Upon seeing a real example, I don't like the clash between () as call operator and comment marker in tips such as

(object, /)
('/' marks preceding arguments as positional-only)
Append object to the end of the list.

I prefer either [] or # as the comment marker.  What think you?

(object, /)
['/' marks preceding arguments as positional-only]
Append object to the end of the list.

(object, /)
`# '/' marks preceding arguments as positional-only`
Append object to the end of the list.

**Review by terryjreedy — CHANGES_REQUESTED** [review]

File: Lib/idlelib/calltips.py
Comment (terryjreedy) [inline Lib/idlelib/calltips.py]: argspec +=  '\n' + _argument_positional

File: Lib/idlelib/idle_test/test_calltips.py
Comment (terryjreedy) [inline Lib/idlelib/idle_test/test_calltips.py:77]: I would prefer that this be truncated also.  We allowed up to 5 docstring lines for builtins to get the signature (bytes needs 5), but if we already have signature, only 1 should be allowed, as below.  However, after checking the code and the docstring, I decided that the source of the problem is that the re.sub docstring is defective in not having an overly long summary sentence and then not having a blank line after the first '.', as it should.  We could check for '.', but not worth the effort now.  Instead I would like to add a comment so it is clear that this case was thought about.  This is a good additional test case.

`# Comment runs on because of long first sentence and missing blank thereafter.  Display as above.`

**Review by terryjreedy — CHANGES_REQUESTED** [review]

File: Lib/idlelib/calltips.py
Comment (mlouielu) [reply Lib/idlelib/calltips.py]: I'll prefer `['/' marks preceding arguments as positional-only]` style, it has more conspicuous when the text was end at the box margin.

Also, I think the word is OK, using google to search will got this [stackoverflow post](https://stackoverflow.com/questions/24735311/python-what-does-the-slash-mean-in-the-output-of-helprange) at the top, it should not be a problem for newbie.

File: Lib/idlelib/calltips.py
Comment (mlouielu) [reply Lib/idlelib/calltips.py]: I'll fix `_argument_positional` to `"\n['/' marks preceding arguments as positional-only]\n"`

**Comment by mlouielu** [comment]
This PR should be block until the regression of not closing tipbox been fixed.

I don't think the problem came from #2306, I test it on mac first `git pr 2822`, then `git revert fae2c3538e`, the problem still exists.

**Comment by mlouielu** [comment]
Block fixed by 89225871d31

File: Lib/idlelib/calltips.py
Comment (terryjreedy) [reply Lib/idlelib/calltips.py]: By 'word', I presume you meant 'positional-only'.  Nice check. We'll go with your preference.

File: Lib/idlelib/calltips.py
Comment (terryjreedy) [reply Lib/idlelib/calltips.py]: Even better.

**Review by terryjreedy — APPROVED** [review]
Requested changes made, some with improvements.

**Pull request merged**
